common/src/main/java/com/linecorp/centraldogma/internal/jsonpatch/DiffProcessor.java

```diff
@@ -0,0 +1,122 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+/*
+ * Copyright (c) 2014, Francis Galiegue (fgaliegue@gmail.com)
+ *
+ * This software is dual-licensed under:
+ *
+ * - the Lesser General Public License (LGPL) version 3.0 or, at your option, any
+ *   later version;
+ * - the Apache Software License (ASL) version 2.0.
+ *
+ * The text of this file and of both licenses is available at the root of this
+ * project or, if you have the jar distribution, in directory META-INF/, under
+ * the names LGPL-3.0.txt and ASL-2.0.txt respectively.
+ *
+ * Direct link to the sources:
+ *
+ * - LGPL 3.0: https://www.gnu.org/licenses/lgpl-3.0.txt
+ * - ASL 2.0: https://www.apache.org/licenses/LICENSE-2.0.txt
+ */
+
+package com.linecorp.centraldogma.internal.jsonpatch;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Supplier;
+
+import javax.annotation.Nullable;
+
+import com.fasterxml.jackson.core.JsonPointer;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.google.common.base.Equivalence;
+import com.google.common.base.Predicate;
+
+// TODO: cleanup
+final class DiffProcessor {
+
+    private static final Equivalence<JsonNode> EQUIVALENCE = JsonNumEquals.getInstance();
+
+    private final List<JsonPatchOperation> diffs = new ArrayList<>();
+    private final ReplaceMode replaceMode;
+    private final Supplier<Map<JsonPointer, JsonNode>> unchangedValuesSupplier;
+
+    DiffProcessor(ReplaceMode replaceMode, final Supplier<Map<JsonPointer, JsonNode>> unchangedValuesSupplier) {
+        this.replaceMode = replaceMode;
+        this.unchangedValuesSupplier = new Supplier<Map<JsonPointer, JsonNode>>() {
+
+            private Map<JsonPointer, JsonNode> unchangedValues;
+
+            @Override
+            public Map<JsonPointer, JsonNode> get() {
+                if (unchangedValues == null) {
+                    unchangedValues = unchangedValuesSupplier.get();
+                }
+                return unchangedValues;
+            }
+        };
+    }
+
+    void valueReplaced(final JsonPointer pointer, final JsonNode oldValue, final JsonNode newValue) {
+        switch (replaceMode) {
+            case RFC6902:
+                diffs.add(new ReplaceOperation(pointer, newValue));
+                break;
+            case SAFE:
+                diffs.add(new SafeReplaceOperation(pointer, oldValue, newValue));
+                break;
+        }
+    }
+
+    void valueRemoved(final JsonPointer pointer) {
+        diffs.add(new RemoveOperation(pointer));
+    }
+
+    void valueAdded(final JsonPointer pointer, final JsonNode value) {
+        final JsonPatchOperation op;
+        if (value.isContainerNode()) {
+            // Use copy operation only for container nodes.
+            final JsonPointer ptr = findUnchangedValue(value);
+            op = ptr != null ? new CopyOperation(ptr, pointer)
```

Comment:
Just curious. Do we need to use `CopyOperation` here? How about using `AddOperation` only without making `unchangedValues` map?

Reply:
Yes, a copy operation generates a smaller patch than an add operation when the value being copied is a container node. Imagine adding vs. copying an array with 100 elements or an object with 100 fields.

Reply:
Ok, thanks! :-)